### PR TITLE
merge

### PR DESCRIPTION
**Pull request by jnwnlee**

### CODE REVIEW

File: fadtk/fad_batch.py [hunk 1]
```diff
@@ -31,8 +31,17 @@ def cache_embedding_files(files: Union[list[Path], str, Path], ml: ModelLoader,
     if isinstance(files, (str, Path)):
         files = list(Path(files).glob('*.*'))
 
+    if force_emb_calc:
```
Comment (keunwoochoi) [inline fadtk/fad_batch.py:34]: add docstring for `force_emb_calc` please

File: fadtk/fad_calculation.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from argparse import ArgumentParser
+
+from .fad import FrechetAudioDistance, log
+from .model_loader import *
```
Comment (keunwoochoi) [inline fadtk/fad_calculation.py:4]: let's avoid using `import *`. seems like it was already used in the original repo but no need to make the same mistake on our end.

File: fadtk/fad_calculation.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+from .fad_batch import cache_embedding_files
+
+def calculate_fad(model_type, baseline, eval, workers=8, force_emb_calc=False, audio_len=4):
+
```
Comment (keunwoochoi) [inline fadtk/fad_calculation.py:8]: docstrings please!

File: fadtk/fad_calculation.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from argparse import ArgumentParser
+
+from .fad import FrechetAudioDistance, log
```
Comment (keunwoochoi) [inline fadtk/fad_calculation.py:3]: we'd better set a logger in this file separately.
```
log = logging.getLogger(__name__)

```

File: fadtk/model_loader.py [hunk 5]
```diff
@@ -54,16 +56,21 @@ def _get_embedding(self, audio: np.ndarray):
     def load_wav(self, wav_file: Path):
         wav_data, _ = soundfile.read(wav_file, dtype='int16')
         wav_data = wav_data / 32768.0  # Convert to [-1.0, +1.0]
-
+        if self.audio_len is not None:
```
Comment (keunwoochoi) [inline fadtk/model_loader.py:59]: can you add comments about what you intend to do here? it'd be very helpful for anyone reading this code in the future, no matter how straightforward this code is (because then they can clarify if your intention is aligned with the implementation or not.)

File: fadtk/model_loader.py [hunk 6]
```diff
@@ -54,16 +56,21 @@ def _get_embedding(self, audio: np.ndarray):
...
+            if wav_data.shape[0] < self.sr*4:
+                padding_size = self.sr*4 - len(wav_data)
+                wav_data = np.pad(wav_data, (0, padding_size), 'constant', constant_values=(0, 0))
```
Comment (keunwoochoi) [inline fadtk/model_loader.py:62]: `wav_data.ndim` be either 1 or 2 depending on the channel number, as we didn't force `always_2d=True`. if stereo files were used, this code might not work.. right? so, i) let's add a comment explaining we're expecting mono file ii) check if it's actually true, and if not -- if the file is stereo -- let's `raise ValueError` here.

File: fadtk/model_loader.py [hunk 7]
```diff
@@ -54,16 +56,21 @@ def _get_embedding(self, audio: np.ndarray):
...
 
 
 class VGGishModel(ModelLoader):
     """
     S. Hershey et al., "CNN Architectures for Large-Scale Audio Classification", ICASSP 2017
     """
-    def __init__(self, use_pca=False, use_activation=False):
-        super().__init__("vggish", 128, 16000)
+    def __init__(self, use_pca=False, use_activation=False, audio_len=None):
+        super().__init__("vggish", 128, 16000, audio_len)
```
Comment (keunwoochoi) [inline fadtk/model_loader.py:73]: docstring please. to explain what `audio_len` does.

File: fadtk/model_loader.py [hunk 8]
```diff
@@ -78,17 +85,110 @@ def load_model(self):
...
+        elif self.variant == 'wavegram-logmel':
+            self.model = panns.Wavegram_Logmel_Cnn14(
+                sample_rate=32000,
+                window_size=1024,
+                hop_size=320,
+                mel_bins=64,
+                fmin=50,
+                fmax=14000,
+                classes_num=527,
+            )
```
Comment (keunwoochoi) [inline fadtk/model_loader.py:165]: add
```
else:
    raise ValueError(f"Unexpected variant: {self.variant}")
```

Comment (keunwoochoi) [reply fadtk/model_loader.py:165]: fyi, because this is stronger and more helpful to the users than adding a type hint.

Comment (keunwoochoi) [reply fadtk/model_loader.py:165]: this comment should go to line 169.

File: fadtk/model_loader.py [hunk 9]
```diff
@@ -78,17 +85,110 @@ def load_model(self):
...
+        return emb
+    
 class EncodecEmbModel(ModelLoader):
     """
     Encodec model from https://github.com/facebookresearch/encodec
 
     Thiss version uses the embedding outputs (continuous values of 128 features).
     """
-    def __init__(self, variant: Literal['48k', '24k'] = '24k'):
+    def __init__(self, variant: Literal['48k', '24k'] = '24k', audio_len=None):
```
Comment (keunwoochoi) [inline fadtk/model_loader.py:189]: ditto

File: fadtk/model_loader.py [hunk 10]
```diff
@@ -264,8 +364,8 @@ class CLAPLaionModel(ModelLoader):
     CLAP model from https://github.com/LAION-AI/CLAP
     """
     
-    def __init__(self, type: Literal['audio', 'music']):
-        super().__init__(f"clap-laion-{type}", 512, 48000)
+    def __init__(self, type: Literal['audio', 'music'], audio_len=None):
```
Comment (keunwoochoi) [inline fadtk/model_loader.py:367]: ditto - on here and everywhere we added `audio_len.

File: fadtk/model_loader.py [hunk 11]
```diff
@@ -613,27 +713,28 @@ def _get_embedding(self, audio: np.ndarray) -> np.ndarray:
 
         return out
 
-
-
-def get_all_models() -> list[ModelLoader]:
+def get_all_models(audio_len=None) -> list[ModelLoader]:
```
Comment (keunwoochoi) [inline fadtk/model_loader.py:716]: if adding the same docstring for `audio_len` is tedious, we can at least add it here.

File: fadtk/panns/finetune_template.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+torch.backends.cudnn.benchmark = True
+torch.manual_seed(0)
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torch.utils.data
+
+from utilities import get_filename
+from models import *
```
Comment (keunwoochoi) [inline fadtk/panns/finetune_template.py:23]: ditto; avoid `import *`.

File: fadtk/panns/losses.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def clip_bce(output_dict, target_dict):
+    """Binary crossentropy loss."""
+    return F.binary_cross_entropy(output_dict["clipwise_output"], target_dict["target"])
+
+
+def get_loss_func(loss_type):
+    if loss_type == "clip_bce":
+        return clip_bce
```
Comment (keunwoochoi) [inline fadtk/panns/losses.py:12]:
```
else:
    raise NotImplementedError(f"Loss type {loss_type} is not implemented.")
```

File: fadtk/panns/pytorch_utils.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
+import numpy as np
+import time
+import torch
+import torch.nn as nn
+
+
+def move_data_to_device(x, device):
+    if "float" in str(x.dtype):
```
Comment (keunwoochoi) [inline fadtk/panns/pytorch_utils.py:8]: so`x` is `np.ndarray`? let's add it to as a comment / docstring.

File: fadtk/panns/pytorch_utils.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+      out: (batch_size, ...)
+    """
+    out = (
+        x[0::2].transpose(0, -1) * mixup_lambda[0::2]
+        + x[1::2].transpose(0, -1) * mixup_lambda[1::2]
+    ).transpose(0, -1)
+    return out
+
+
+def append_to_dict(dict, key, value):
```
Comment (keunwoochoi) [inline fadtk/panns/pytorch_utils.py:36]: optional: consider using `defaultdict`.

File: fadtk/panns/pytorch_utils.py [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+
+    Returns:
+      audio_name: (audios_num,)
+      clipwise_output: (audios_num, classes_num)
+      (ifexist) segmentwise_output: (audios_num, segments_num, classes_num)
+      (ifexist) framewise_output: (audios_num, frames_num, classes_num)
+      (optional) return_input: (audios_num, segment_samples)
+      (optional) return_target: (audios_num, classes_num)
+    """
+    output_dict = {}
```
Comment (keunwoochoi) [inline fadtk/panns/pytorch_utils.py:60]: this is the part we can use `defaultdict(list)`

File: fadtk/panns/pytorch_utils.py [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+            time1 = time.time()
+
+    for key in output_dict.keys():
+        output_dict[key] = np.concatenate(output_dict[key], axis=0)
+
+    return output_dict
+
+
+def interpolate(x, ratio):
+    """Interpolate data in time domain. This is used to compensate the
```
Comment (keunwoochoi) [inline fadtk/panns/pytorch_utils.py:117]: the function name is interpolate, but it actually doesn't perform interpolation. assuming you implemented what you intended, something like `upsample_temporal_dim` would sound more accurate to me.

File: fadtk/panns/utilities.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+import logging
+import numpy as np
+from scipy import stats
+import datetime
+import pickle
+
+
+def create_folder(fd):
+    if not os.path.exists(fd):
+        os.makedirs(fd)
```
Comment (keunwoochoi) [inline fadtk/panns/utilities.py:11]: `os.makedirs(fd, exist_ok=True)` would do the job.

File: fadtk/panns/utilities.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+        return x[0:audio_length]
+
+
+def d_prime(auc):
+    d_prime = stats.norm().ppf(auc) * np.sqrt(2.0)
+    return d_prime
+
+
+class Mixup(object):
+    def __init__(self, mixup_alpha, random_seed=1234):
```
Comment (keunwoochoi) [inline fadtk/panns/utilities.py:118]: do we want fixed random seed? i don't know how this class is used but this can be dangerous(ly duplicate the mixup process for all the workers).

File: fadtk/fad_calculation.py (hunk 4 above)
Comment (jnwnlee) [reply fadtk/fad_calculation.py:3]: 'fad_calculation.py' seems to be added by Modan, but not used now.
I'll remove the file. Sorry for the confusion.

File: fadtk/model_loader.py (hunk 6 above)
Comment (jnwnlee) [reply fadtk/model_loader.py:62]: Instead of chopping and padding, I suggest to modify the role of audio_len: to check if the actual audio length matches the given.
I'll update the docstrings and implementation.

File: fadtk/panns/pytorch_utils.py (hunk 17 above)
Comment (jnwnlee) [reply fadtk/panns/pytorch_utils.py:117]: I changed the function name. Thanks for the suggestion!

File: fadtk/panns/utilities.py (hunk 19 above)
Comment (jnwnlee) [reply fadtk/panns/utilities.py:118]: It seems we are not using this function.
TBH, I brought the required codes file-wise from the original repo. There may be some unnecessary parts. I'll clean soon.
